plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -590,27 +595,158 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String bucketName = getBucketName(uri);
+        String key = keyFromPath(path);
+        ObjectMetadata s3ObjectMetadata = getS3ObjectMetadata(path, bucketName, key);
+        ObjectMetadata s3DirectoryMetadata = null;
+        if (!key.isEmpty()) {
+            s3DirectoryMetadata = getS3ObjectMetadata(path, bucketName, key + PATH_SEPARATOR);
+        }
+
+        if (recursive) {
+            boolean deleted = false;
+            if (s3ObjectMetadata == null ||
+                    s3DirectoryMetadata == null ||
+                    MediaType.parse(s3DirectoryMetadata.getContentType()).is(DIRECTORY_MEDIA_TYPE)) {
+                deleted = deletePrefix(path);
+            }
+            if (key.endsWith(DIRECTORY_SUFFIX)) {
+                Path strippedPath = new Path(stripDirectorySuffix(path.toString()));
+                deleted = deletePrefix(strippedPath) || deleted;
             }
+            if (s3ObjectMetadata != null) {
+                deleted = deleteObject(key) || deleted;
+            }
+            if (s3DirectoryMetadata != null) {
+                if (MediaType.parse(s3DirectoryMetadata.getContentType()).is(DIRECTORY_MEDIA_TYPE) ||
+                        key.endsWith(DIRECTORY_SUFFIX)) {
+                    String directoryKey = (key.endsWith(DIRECTORY_SUFFIX) ? stripDirectorySuffix(key) : key) + PATH_SEPARATOR;
+                    deleted = deleteObject(directoryKey) || deleted;
+                }
+            }
+            return deleted;
         }
-        catch (FileNotFoundException e) {
+        else {
+            if (s3ObjectMetadata == null && s3DirectoryMetadata == null) {
+                if (path.getName().isEmpty()) {
+                    return false; // bucket root does not exist
+                }
+                if (key.endsWith(DIRECTORY_SUFFIX)) {
+                    Path strippedPath = new Path(stripDirectorySuffix(path.toString()));
```

Comment:
I'm rather reluctant to do the stripping of the directory suffix, but it seems that S3 doesn't list anything under the prefix, when the prefix contains  the `_$folder$` placeholder.

cc @pettyjamesm  any ideas how to handle this use case more elegantly?

Reply:
Oof, you're right. Ok, so... let's say that we have a delete for `s3://bucket/root_$folder$`. It stands to reason that we don't need to do anything special for recursive deletes underneath the root (eg: `s3://bucket/root/subpath_$folder$`) but at the top level there's a question about how to proceed. I think it stands to reason that for the top-level delete case, we want to strip `_$folder$` both when checking for nested "files or directories"  (ie: fail to delete if non-empty and non-recursive) as well as when doing a recursive delete (ie: delete `s3://bucket/root_$folder$` as well as object prefixed by `s3://bucket/root/`).